### ISSUE TITLE
Review architecture, models, and design of semantic search system

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#4962]
We have done a lot of work towards our semantic search engine, including selecting models, designing the architecture, and building a microservice (named inception), to do the embedding process. 

Rachel is our new AI engineer, so she needs to get up to speed on all this. This ticket is for her to do that, ask questions about what she sees, and identify anything she'd suggest that we might want to do differently. (Hopefully that's not much!).

**Comment by rachlllg** [comment]
Using this space to summarize all the experiments and findings so far:

- [Semantic search backlog](https://github.com/orgs/freelawproject/projects/56/views/1) 
- Overall evaluation and experimental findings: freelawproject/courtlistener/issues/4277
   - Conclusion: 
      - Experimental objective: create a semantic search in the form of question-answering based on embeddings generated using an open source model
      - Experimental dataset: 10 randomly sampled SCOTUS opinions, with "query" generated using llamaindex to form query-opinion pairs for finetuning and evaluation.
      - Model selected: [sentence-transformers/multi-qa-mpnet-base-dot-v1](https://huggingface.co/sentence-transformers/multi-qa-mpnet-base-dot-v1)
- Vector DB options: freelawproject/courtlistener/issues/3489
   - Conclusion: 
      - Use S3 as the overall database, with self-hosted Elastic Search to perform [vector search](https://www.elastic.co/what-is/vector-search)
- Discussion & the architecture of the micro service: freelawproject/courtlistener/issues/4530
- [Our open source micro service - Inception](https://github.com/freelawproject/inception/tree/main)

Questions/Concerns @mlissner 
1. Task formation:
   - Do we intend the search engine to be in a question-answering format (ie, the user asks a question and we retrieve opinions based on the question)? Because the model we chose is optimized for question-answering, one concern is:
      - Correct me if I'm wrong, we do not have a dataset that contains the actual questions a user might ask and the actual documents that "best" match the answers, in another word, we do not have a golden standard test set that we can use to objectively evaluate the performance. I think previous experiments used llamaindex to generate question-document pairs, my concern is the user question could be drastically different from what llamaindex generates (ie, overly verbose user questions could impact retrieval performance), hence, production results could look very different from experiments.
      -I wonder if it might be a slightly easier (more readily attainable task) to form this as a similarity-retrieval task. In another word, instead of searching by a question, we search by a statement. The main difference is, a model that's optimized for question-answering was trained with "question-answering" pairs, thus expects a question, but when we don't have these pairs, it becomes hard to finetune the models and evaluate the performance. Whereas a model optimized for similarity-retrieval doesn't require a question, it only requires the texts (which we already have). I will poke around with this option to see if there is any meat here. Also want to try the sentence_transformer tagged on top of ModernBert to take advantage of the enhanced context window (~8k tokens) to generate potentially one embedding per opinion (or a handful for really long ones).
2. Finetuning:
   - Can you remind me why we decided against using a finetuned model? Is the lack of golden standards one of the roadblocks?
3. Reranking:
   - Can you also remind me why we decided against reranking? Is the native results already good enough so reranking is more of a nice-to-have?
4. Hybrid Search:
   - Is the current plan for user to toggle between key-word search and semantic search? With maybe future implementation to incorporate hybrid search?
5. Test data:
   - I noticed our experiments were conducted with only SCOTUS data, I know we have a lot of data from other jurisdictions, my concern is SCOTUS data is perhaps not representative of our population, ie, there might be some sampling bias here that inflates the actual performance. I plan to get some non-SCOTUS data to test the robustness of the performance.

**Comment by mlissner** [comment]
And we're off to the races. Thanks Rachel. 

A few thoughts, but hopefully @legaltextai will chime in too, if he has a minute...

> Finetuning & Reranking

IIRC, we pushed these out because we wanted to just get something out the door, and they felt like something we could add later. There's no reason we have to do it now, right? 

> Hybrid search

Right now, we have heavily faceted keyword search. The idea is to give users an option when they do their query so that it is either interpreted as a keyword search or as semantic search. (Imagine a checkbox for "do semantic search" or something.) Sergei also wants to use AI to do this automatically, but that's for down the road. 

Assuming that users can toggle between semantic and keyword, the hope is that the existing facets (which I believe is the same thing as hybrid), can be used along with whatever type of query is performed. So either of these should work:

```
Court: SCOTUS
Date: After 2015-01-01
Query: intellectual property
Query Type: Semantic
```

```
Court: SCOTUS
Date: After 2015-01-01
Query: intellectual property
Query Type: keyword
```

> my concern is SCOTUS data is perhaps not representative of our population, ie, there might be some sampling bias here that inflates the actual performance

Totally possible.

> we do not have a dataset that contains the actual questions a user might ask and the actual documents that "best" match the answers

You are correct. We have case law and we have filings.

> Do we intend the search engine to be in a question-answering format

I think that'd be part of it and we should perform well when that's the case, but I guess some people will put in queries like that and others will put in queries that look more like keyword search? We have about 10M user queries that can be analyzed (of course, they're based on our current UI). It might not be crazy to do a quick survey of other systems and see what kinds of queries they're asking users to do. I think Sergei encouraged questions in his demos. 

> I will poke around with this option to see if there is any meat here.

Sounds good, please do.

> Also want to try the sentence_transformer tagged on top of ModernBert to take advantage of the enhanced context window (~8k tokens) to generate potentially one embedding per opinion (or a handful for really long ones).

I think we found that longer context led to worse results, IIRC, but I'm not sure. One thing to remember is that longer decisions cover multiple topics, so breaking that into smaller windows doesn't always feel like a crazy idea. Our longest decision is 1,000 pages. 

Thanks for the analysis.

**Comment by rachlllg** [comment]
> Finetuning & Reranking

> IIRC, we pushed these out because we wanted to just get something out the door, and they felt like something we could add later. There's no reason we have to do it now, right?

Agree, we can do these in future iterations. 
- In the short term, we can consider adding reranking using an existing open source framework (need more time to compare, can get to it after citator). 
- In the long term, hopefully we will have a good dataset with actual user queries and golden documents that match with these queries that we can use to finetune a model (which would most likely lead to improved performance). The finetuned model can be pushed to huggingface hub and we can load it similar to how we load other huggingface models. The embeddings would need to be regenerated using the finetuned model.

> I think we found that longer context led to worse results, IIRC, but I'm not sure. One thing to remember is that longer decisions cover multiple topics, so breaking that into smaller windows doesn't always feel like a crazy idea. Our longest decision is 1,000 pages.

From my experiments (see below results), I found that when the model supports long context windows, utilizing longer context lead to improved performance. I think this makes sense: consider when we are reading a story, there may be important facts introduced early on being mentioned later in the story, by chunking the story to (effectively) short paragraphs, the later paragraphs are lacking these important prior facts, longer context window solves this problem and produces more contextually rich embeddings. Majority of opinions in Case Law are less than 3000 words in length, using a model with longer context size would allow us to capture the semantics from the entire opinion, creating a complete picture of each opinion.

Below is the conclusion of my analysis based on my experiments. More details on the dataset, experimental design, limitation & future works can be found at [this repo](https://github.com/rachlllg/FLP_Semantic_Search) (let me know if there is a better place to house this repo within FLP, happy to move it).

The two models I recommend are:

- `multi-qa-mpnet-base-dot-v1` with 512 chunk size
- `nomic-ai/modernbert-embed-base` with 8192 chunk size

With a slight preference towards `nomic-ai/modernbert-embed-base` with 8192 chunk size, given:
- It provides a five percentage point advantage in both metrics without a substantial increase in latency
- It is finetuned on the latest encoder SOTA model (ModernBERT)
- It allows for efficient long context embedding and retrieval, which is better suited for longer corpus such as legal opinions

However, I know we already have work in place to use `multi-qa-mpnet-base-dot-v1`, which I think is a good starting model for v1 of semantic search, we can consider upgrading to `nomic-ai/modernbert-embed-base` with 8192 chunk size in v2 together with finetuning. However, if we use `multi-qa-mpnet-base-dot-v1`, I recommend using 512 as the chunk size, instead of the 350 currently in place in Inception. We can also consider adding logic for sentence overlap between chunks.

Below are the experiment results, note the latency here (seconds per query) includes loading the model, generating the embeddings for the query and the opinions, and retrieval, the latency in production should be much less as the model would be cached and the opinion embeddings are pre-generated:
<img width="1151" alt="Image" src="https://github.com/user-attachments/assets/fdbe0cba-2934-40cd-b0ad-49b8c45d5164" />

<img width="1148" alt="Image" src="https://github.com/user-attachments/assets/807c3f54-3231-48b9-aafd-bf6e88f6cbde" />

<img width="1153" alt="Image" src="https://github.com/user-attachments/assets/79e860df-1f94-4e05-815d-c63a638be696" />

**Comment by legaltextai** [comment]
great work, @rachlllg. i don't have many comments but i would encourage to finalize the embedding model first as vectorizing will take a long time and you may not want to re-run it with a different model any time soon. 

350 is a number of words , if i am not mistaken, which should roughly translate to 512 tokens. 

also, think of possible query transformation as a way to improve search results. [Hyde](https://docs.llamaindex.ai/en/stable/examples/query_transformations/HyDEQueryTransformDemo/) looks promising and some other options incl from langchain. CL has been logging user queries for some time , those will be useful to learn from.

**Comment by rachlllg** [comment]
Thanks @legaltextai ! And thank you for your guidance and valuable ideas/discussions!

> 350 is a number of words , if i am not mistaken, which should roughly translate to 512 tokens.

You are right, the implementation currently is by counting words, I wonder if we can use AutoTokenizer to tokenize so we ensure the texts are chunked to at most 512 tokens? Some longer words may translate to multiple tokens, resulting in words being truncated if exceeding 512 tokens. But perhaps that is too complex and this current implementation is already good to get things started.

> also, think of possible query transformation as a way to improve search results. [Hyde](https://docs.llamaindex.ai/en/stable/examples/query_transformations/HyDEQueryTransformDemo/) looks promising and some other options incl from langchain. CL has been logging user queries for some time , those will be useful to learn from.

[HyDE](https://arxiv.org/abs/2212.10496) relies on a decoder LLM (a different LLM from the encoder LLM we use to generate the embeddings) to transform the queries, the risk would be potential hallucination from this added LLM, and it would ideally be tuned to legal context. Given we don't want to pass user queries to outside APIs, we would need to use an open source decoder LLM (vs sota decoders like gpt or claude). I think it would be a great idea for future improvements, I added to the documentation, thank you for the suggestion!

**Comment by mlissner** [comment]
> the latency here (seconds per query) includes loading the model, generating the embeddings for the query and the opinions, and retrieval

I'm finding it hard to make sense of what this will look like in production. If it's 0.6s including all of the above, how fast is a vectorizing a query? Or maybe we don't care and we just have to suck it up no matter what — if the latency is slightly worse for something, but the search performance is better, perhaps we just don't care about the latency. (They do seem pretty similar anyway.)

> can use AutoTokenizer

I don't know what autotokenizer is, but I can guess that it's some Python thing that does a nicer job converting words to tokens? Assuming that's right, how long would it take to put this into use in inception?

**Comment by rachlllg** [comment]
> I'm finding it hard to make sense of what this will look like in production. If it's 0.6s including all of the above, how fast is a vectorizing a query? Or maybe we don't care and we just have to suck it up no matter what — if the latency is slightly worse for something, but the search performance is better, perhaps we just don't care about the latency. (They do seem pretty similar anyway.)

Yea I apologize this latency metric is a bit "flawed", I wanted to get a sense of how fast it is to run each model from beginning to end (from raw documents to evaluation), which is not what we would do in production. In my test, on my local CPU (Mac M2 chip with 8 core), it took ~14s to chunk and embed 450 queries using `nomic-ai/modernbert-embed-base`, so that's 0.03s per query, and that's without any parallel processing and with a ton of other stuff running at the same time.
